### ISSUE TITLE
Support fetching sources from spec + sources file to build SRPM

### ISSUE THREAD

**Issue by qsuscs** [fedora-copr/copr#3044]
Currently, COPR supports three methods to build an SRPM from SCM sources: rpkg[^1], tito (and tito test), and `make srpm`.  For a repository built like the Fedora standard (spec file with URLs in the `SourceN` fields, sources file with checksums), there is no direct way to let COPR handle the SRPM.

While rpkg supports getting the source files and building SRPMs, it can only fetch from DistGit, not upstream.  COPR uses DistGit internally, and this works for subsequent builds, but users do not seem to have write access there and thus cannot use e.&#8239;g.&nbsp;`fedpkg new-sources`.

I still have not understood if tito is suitable for packaging an “external project”, i.&#8239;e., the maintainer of the RPM and upstream are different entities and the packaging repository is distinct.  But again, not COPR’s fault.

This leaves `make srpm`.  I have done this in the past, with the Makefile below as `.copr/Makefile`, invoking `spectool` manually to download the sources, `sha512sum` to verify them, and `rpmbuild` to build the SRPM.  For something I would consider a standard and trivial case, this is needlessly complicated.

I do not (yet) maintain any package in Fedora itself, so I do not know the process there exactly, but from the documentation, it seems like COPR works differently in this aspect.  This is annoying and confusing, especially when one is planning to submit a package to Fedora eventually and uses COPR as a sort of test environment.

**TL;DR:** Please add support to automatically fetch source tarballs and build SRPMs from spec files with `SourceN: https://…` lines and `sources` files with checksums (like in Fedora package Git repositories) and/or allow backend DistGit access to the user.

---

```Makefile
spec ?= btop.spec
outdir ?= $(CURDIR)
sourcedir = $(shell rpmbuild -E '%_sourcedir')

srpm:
	dnf -y install rpmdevtools
	spectool -R -g $(spec)
	env -C $(sourcedir) \
		sha512sum -c $(CURDIR)/sources || \
		(find $(sourcedir) -type f -delete; exit 1)
	rpmbuild -D "_srcrpmdir $(outdir)" -bs $(spec)
```
[^1]: Wrong place to complain about this, but it was and is really confusing for me that rpkg ≠ rpkg-util, that COPR uses the latter (I guess?) but calls it rpkg, and that it is “no longer developed” as per https://pagure.io/rpkg-util.

**Comment by FrostyX** [comment]
> TL;DR: Please add support to automatically fetch source tarballs and build SRPMs from spec files with SourceN: https://… lines and sources files with checksums (like in Fedora package Git repositories) and/or allow backend DistGit access to the user.

@qsuscs If I understand correctly, there may be a simple workaround. [Submitting a build via URL](https://docs.pagure.org/copr.copr/user_documentation.html#urls) allows not only linking SRPM also a specfile. When all `Source`s are URLs, then they are automatically downloaded.

Can you maybe use the RAW URL for your spec file in your github/gitlab/pagure repository?

**Comment by azrdev** [comment]
> When all `Source`s are URLs, then they are automatically downloaded.
> 
> Can you maybe use the RAW URL for your spec file in your github/gitlab/pagure repository?

That would require putting URLs as Source:, correct? sounds inelegant to me, but might work

**Comment by FrostyX** [comment]
I got an email notification from GitHub about @azrdev replying to this conversation but no comment here. I don't understand why.

>> When all Sources are URLs, then they are automatically downloaded.
>> Can you maybe use the RAW URL for your spec file in your github/gitlab/pagure repository?
>
> @azrdev - That would require putting URLs as Source:, correct? sounds inelegant to me, but might work

I think you are correct. To be sure, let me give an example. Something like:

```
Source0:    https://github.com/haikarainen/%{name}/archive/v%{version}/%{name}-%{version}.tar.gz
```

or 

```
%{URL}/archive/v%{version}/%{name}-%{version}.tar.gz
```

For example, here is one of my packages:
https://src.fedoraproject.org/rpms/light/raw/rawhide/f/light.spec
I can use URL for this specfile, submit it to Copr, and it will build successfully. No SRPM is needed.

**Comment by azrdev** [comment]
> I got an email notification from GitHub about @azrdev replying to this conversation but no comment here. I don't understand why.

I think I removed my comment when I noticed the original issue already answered the question with the line
> spec file with URLs in the SourceN fields



> AFAIK it is a standard practise to include the full URL

I agree with you concerning upstream sources, but I was thinking about patches or other files included with the spec file, not upstream.

**Comment by azrdev** [comment]
unfortunately this is still an issue:

having a packaging repo (e.g.  <https://github.com/azrdev/monaspace-rpm>) with .spec file and source/patch files, and upstream sources as `Source:` in the spec -- trying to build this on copr with "Source Type: Build from an SCM repository" and "Build SRPM with: rpkg" successfully downloads the upstream source, but fails to use the sources shipped with the spec. builder-live.log.gz excerpt from copr:

```
RPM build errors:
stderr: warning: Downloading https://github.com/githubnext/monaspace/archive/v1.000/monaspace-v1.000.tar.gz to /var/lib/copr-rpmbuild/results/monaspace-v1.000.tar.gz
Building target platforms: noarch
Building for target noarch
error: Bad file: /var/lib/copr-rpmbuild/results/69-monaspace-neon-fonts.conf: No such file or directory
error: Bad file: /var/lib/copr-rpmbuild/results/69-monaspace-argon-fonts.conf: No such file or directory
error: Bad file: /var/lib/copr-rpmbuild/results/69-monaspace-xenon-fonts.conf: No such file or directory
error: Bad file: /var/lib/copr-rpmbuild/results/69-monaspace-radon-fonts.conf: No such file or directory
error: Bad file: /var/lib/copr-rpmbuild/results/69-monaspace-krypton-fonts.conf: No such file or directory
```
Looking in the COPR docs yields https://docs.pagure.org/copr.copr/user_documentation.html#weird-scm-build-failure which hints that this used to work before rpkg-util v2 was replaced by rpkg v3, but goes on explaining how a different invocation of rpkg would fix this, which is not possible via COPR.

There really is lack of user documentation for this use case.

**Comment by FrostyX** [comment]
Hello @azrdev,
I don't use `rpkg` for this purpose, so I am not 100% sure with my answer, but I think I made it work for your package. I think the trick was using fully qualified URLs.

See my changes here:
https://github.com/FrostyX/monaspace-rpm/commits/main/

and the build succeeded:
https://copr.fedorainfracloud.org/coprs/frostyx/copr-rpkg-issue-3044/build/10121991/

I am not sure if this is what you were trying to achieve, or if I did some weird shortcut that doesn't actually solve the issue. Please let me know. If this works as expected, we could update the documentation you linked.

**Comment by azrdev** [comment]
> the trick was using fully qualified URLs.

yes, this fixes the build, but 1) hardcodes the path to the packaging repo, which is not necessary otherwise 
2) AIUI it is normal to have "local" source files (next to the .spec) referenced by relative path

**Comment by FrostyX** [comment]
Fair enough. I was reading through the documentation and figured out this:

https://github.com/FrostyX/monaspace-rpm/commit/aa5d3329434778d26e6465e0fa4cfb984b9609ed
https://copr.fedorainfracloud.org/coprs/frostyx/copr-rpkg-issue-3044/build/10161456/

Maybe that could work?
